napari/components/_tests/test_layer_slicer.py

```diff
@@ -0,0 +1,264 @@
+
+
+def test_slice_layers_async_lock_blocking(layer_slicer):
+    dims = Dims()
+    layer = FakeAsyncLayer()
+
+    assert layer.slice_count == 0
+    with layer.lock:
+        blocked = layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        assert not blocked.done()
+
+    assert layer.slice_count == 1
+    assert blocked.result()[layer].id == 1
+
+
+def test_slice_layers_async_multiple_calls_cancels_pending(layer_slicer):
+    dims = Dims()
+    layer = FakeAsyncLayer()
+
+    with layer.lock:
+        blocked = layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        pending = layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        assert not pending.running()
+        assert pending._state == 'PENDING'
+        layer_slicer.slice_layers_async(layers=[layer], dims=dims)
+        assert not blocked.done()
+
+    assert pending.cancelled()
+
+
+def test_slice_layers_mixed_allows_sync_to_run(layer_slicer):
+    """ensure that a blocked async slice doesn't block sync slicing"""
+    dims = Dims()
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeSyncLayer()
+    with layer1.lock:
+        blocked = layer_slicer.slice_layers_async(layers=[layer1], dims=dims)
+        layer_slicer.slice_layers_async(layers=[layer2], dims=dims)
+        assert layer2.slice_count == 1
+        assert not blocked.done()
+
+    assert blocked.result()[layer1].id == 1
+
+
+def test_slice_layers_mixed_allows_sync_to_run_one_slicer_call(layer_slicer):
+    """ensure that a blocked async slice doesn't block sync slicing"""
+    dims = Dims()
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeSyncLayer()
+    with layer1.lock:
+        blocked = layer_slicer.slice_layers_async(
+            layers=[layer1, layer2], dims=dims
+        )
+
+        assert layer2.slice_count == 1
+        assert not blocked.done()
+
+    assert blocked.result()[layer1].id == 1
+
+
+def test_slice_layers_async_with_multiple_async_layer_with_all_locked(
+    layer_slicer,
+):
+    """ensure that if only one layer has a lock, the nonlocked layer can continue"""
+    dims = Dims()
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeAsyncLayer()
+
+    with layer1.lock, layer2.lock:
+        blocked = layer_slicer.slice_layers_async(
+            layers=[layer1, layer2], dims=dims
+        )
+
+        assert not blocked.done()
+
+    assert blocked.result()[layer1].id == 1
+    assert blocked.result()[layer2].id == 1
+
+
+def test_slice_layers_async_with_multiple_async_layer_with_one_lock(
+    layer_slicer,
+):
+    """ensure that if only one layer has a lock, the nonlocked layer can continue"""
+    dims = Dims()
+    layer1 = FakeAsyncLayer()
+    layer2 = FakeAsyncLayer()
+
+    with layer1.lock:
+        blocked = layer_slicer.slice_layers_async(
+            layers=[layer1, layer2], dims=dims
+        )
+        # TODO: kcp: still wish we had finer control here to ensure that layer2
+        #       is not blocked, but layer1 is
```

Comment:
Would this be enough?

```suggestion
        assert layer2.slice_count == 1
```

Given the way we setup the tests and fakes, we know that means slicing layer2 finished.